labgrid/protocol/commandprotocol.py

```diff
@@ -1,6 +1,87 @@
 import abc
 
 
+class CommandProcessProtocol(abc.ABC):
+    """Abstract class for a running command"""
+
+    @abc.abstractmethod
+    def read(self, size=-1, *, timeout=None):
+        """
+        Reads bytes from the process until either size bytes have been read,
+        timeout seconds have elapsed, or EOF is encountered. Returns as many
+        bytes as were read until that happens.
+
+        If size is -1, read as much data as possible until either the timeout
+        or EOF is encountered.
+
+        If timeout is None, there is no timeout and reading will continue until
+        either size bytes have been read, or EOF occurs. A negative timeout
+        means to use the default timeout for this process
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def read_nonblocking(self, size=1, timeout=-1):
+        """
+        This reads at most size characters from the process, or until EOF or a
+        timeout occurs. If no data is read before EOF or a timeout occurs, then
+        pexpect.EOF or pexpect.TIMEOUT are raised respectively.
+
+        A timeout of None means to wait forever for either data or EOF
+
+        Note that the timer is per-character, not for the entire buffer
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def write(self, data):
+        """
+        Write data to the process
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def poll(self):
+        """
+        Check if the process is alive. If the process is alive, None will be
+        returned. If the process exited normally, the return code will be
+        returned. If the process died from a signal, the negative value of the
+        signal number will be returned.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def stop(self):
```

Comment:
nice work :)
I tested it with my setup and run into a issue while running:
```
def test_shell_interrupt(shell_command):
    with shell_command.popen("yes") as p:
        l = p.read(1024).decode("utf-8").splitlines()
        assert l == ["y"] * len(l)

    shell_command.run_check("true")
```

on the console log I see:
```
root@ksp0627-imx6-1:~# MARKER='XVUS''LJFJHA' run yes
XVUSLJFJHAy
y
y


y
y
^C
root@ksp0627-imx6-1:~#
```

put the test runs into a 60sec  timout
```

        SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!YXLELHEROR)(?!YXLELHERO$)(?!YXLELHER$)(?!YXLELHE$)(?!YXLELH$)(?!YXLEL$)(?!YXLE$)(?!YXL$)(?!YX$)(?!Y$).', re.DOTALL), re.compile(b'^YXLELHEROR\\s+(\\d+)\\s+.*root@[\\w-]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}




       SerialDriver.expect state='stop' result=(2, b'', <class 'pexpect.exceptions.TIMEOUT'>, <class 'pexpect.exceptions.TIMEOUT'>) duration='60.007'
     ConsoleMarkerProcess.expect state='stop' result=(1, b'\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\n^C\r\n', <re.Match object; span=(711, 733), match=b'root@ksp0627-imx6-1:~#'>, b'root@ksp0627-imx6-1:~#') duration='60.087'
   ConsoleMarkerProcess.stop state='stop' duration='60.088'
```

Seams `read_nonblocking` not give the "promt" only case without a timout for busybox shells

Reply:
@jremmet Hmm, that suspiciously looks like the problem I fixed yesterday with busybox.... can you please make sure verify you have the latest commits I pushed? If you do please let me know and I'll look more; specifically, there should a additional regex that matches the prompt in the `SerialDriver.expect` (for a total of 3 regex it's looking for)

Reply:
Thanks for the fast response. I'm on 83c64e9 JoshuaWatt/labgrid.git@command-background. The timed out expext seams to come from `ConsoleMarkerProcess` `read_nonblocking`. There I found no `_prompt_re`.

Reply:
Huh, I'm not sure what is going on here; I've been testing with busybox and can't seem to reproduce what's going on here. The weird part is that `stop` is waiting 60 seconds but then returning that if found the prompt anyway

Can you do a little more digging and provides some more logs?

Reply:
After updating to 58d847d to timeout is gone :+1: